File: DexieNET/DexieNET/Cloud/DexieNETCloud.cs
/*
DexieNETCloud.cs

Copyright(c) 2023 Bernhard Straub

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

'DexieNET' used with permission of David Fahlander 
*/

using Microsoft.JSInterop;

namespace DexieNET
{
    public interface IDBCloudEntity
    {
        string? Owner { get; }
        string? RealmId { get; }

        public string? EntityKey => GetEntityKey();

        private string? GetEntityKey()
        {
            if (Owner is null || RealmId is null)
            {
                return null;
            }

            return Owner + RealmId;
        }
    }

    public record DBCloudEntity(string Owner, string RealmId) : IDBCloudEntity
    {
        public string EntityKey => Owner + RealmId;
    }

    public static partial class DBCloudExtensions
    {
        public static void ConfigureCloud(this DBBase dexie, DexieCloudOptions cloudOptions)
        {
            if (!dexie.CloudSync)
            {
                throw new InvalidOperationException("Can not ConfigureCloud for non cloud database.");
            }

            if (cloudOptions.UnsyncedTables is null)
            {
                cloudOptions = cloudOptions
                    .WithUnsyncedTables(dexie.UnsyncedTables);
            }

            var jsi = cloudOptions.FromObject();
            var err = dexie.DBBaseJS.Module.Invoke<string?>("ConfigureCloud", dexie.DBBaseJS.Reference, jsi);

            if (err is not null)
            {
                throw new InvalidOperationException(err);
            }
        }

        #region Sharing
        public static string GetTiedRealmID(this DBBase dexie, string id)
        {
            if (!dexie.CloudSync)
            {
                throw new InvalidOperationException("Can not call GetTiedRealmID for non cloud database.");
            }

            return dexie.DBBaseJS.Module.Invoke<string>("GetTiedRealmID", id);
        }
        #endregion

        #region UIInteraction
        public static IObservable<UIInteraction> UserInteractionObservable(this DBBase dexie)
        {
            dexie.TestCloudSync();
            var jso = new JSObservable<UIInteraction>(dexie, "SubscribeUserInteraction", "ClearUserInteraction");
            return jso;
        }

        public static void SubmitUserInteraction(this DBBase dexie, UIInteraction interaction, UIParam param)
        {
            dexie.TestCloudSync();
            dexie.DBBaseJS.Module.InvokeVoid("OnSubmitUserInteraction", interaction.Key, param.AsDictionary());
        }

        public static void SubmitUserInteraction(this DBBase dexie, UIInteraction interaction, Dictionary<string, string> param)
        {
            dexie.TestCloudSync();
            dexie.DBBaseJS.Module.InvokeVoid("OnSubmitUserInteraction", interaction.Key, param);
        }

        public static void CancelUserInteraction(this DBBase dexie, UIInteraction interaction)
        {
            dexie.TestCloudSync();
            dexie.DBBaseJS.Module.InvokeVoid("OnCancelUserInteraction", interaction.Key);
        }
        #endregion

        #region Invites
        public static IObservable<IEnumerable<Invite>> InvitesObservable(this DBBase dexie)
        {
            dexie.TestCloudSync();
            var jso = new JSObservable<IEnumerable<Invite>>(dexie, "SubscribeInvites", "ClearInvites");
            return jso;
        }

        public static void AcceptInvite(this DBBase dexie, Invite invite)
        {
            dexie.TestCloudSync();
            dexie.DBBaseJS.Module.InvokeVoid("AcceptInvite", invite.Id);
        }

        public static async ValueTask AcceptInvite(this DBBase dexie, Member member)
        {
            dexie.TestCloudSync();
            await dexie.DBBaseJS.Module.InvokeVoidAsync("AcceptInviteMember", dexie.DBBaseJS.Reference, member.Id);
        }

        public static void RejectInvite(this DBBase dexie, Invite invite)
        {
            dexie.TestCloudSync();
            dexie.DBBaseJS.Module.InvokeVoid("RejectInvite", invite.Id);
        }

        public static async ValueTask RejectInvite(this DBBase dexie, Member member)
        {
            dexie.TestCloudSync();
            await dexie.DBBaseJS.Module.InvokeVoidAsync("RejectInviteMember", dexie.DBBaseJS.Reference, member.Id);
        }

        public static async ValueTask ClearInvite(this DBBase dexie, Member member)
        {
            dexie.TestCloudSync();
            await dexie.DBBaseJS.Module.InvokeVoidAsync("ClearInviteMember", dexie.DBBaseJS.Reference, member.Id);
        }
        #endregion

        #region Roles
        public static IObservable<Dictionary<string, Role>> RoleObservable(this DBBase dexie)
        {
            dexie.TestCloudSync();
            return new JSObservable<Dictionary<string, Role>>(dexie, "SubscribeRoles");
        }
        #endregion

        #region SyncState
        public static IObservable<SyncState> SyncStateObservable(this DBBase dexie)
        {
            dexie.TestCloudSync();
            return new JSObservable<SyncState>(dexie, "SubscribeSyncState");
        }
        #endregion

        #region UserLogin
        public static IObservable<UserLogin> UserLoginObservable(this DBBase dexie)
        {
            dexie.TestCloudSync();
            return new JSObservable<UserLogin>(dexie, "SubscribeUserLogin");
        }

        public static string CurrentUserId(this DBBase dexie)
        {
            dexie.TestCloudSync();
            return dexie.DBBaseJS.Module.Invoke<string>("CurrentUserId", dexie.DBBaseJS.Reference);
        }

        public static async ValueTask UserLogin(this DBBase dexie, LoginInformation userLogin)
        {
            dexie.TestCloudSync();
            var grantType = userLogin.GrantType switch
            {
                GrantType.DEMO => "demo",
                GrantType.OTP => "otp",
                _ => throw new InvalidOperationException("GrantType: Invalid type!")
            };

            await dexie.DBBaseJS.Module.InvokeVoidAsync("UserLogin", dexie.DBBaseJS.Reference, userLogin.EMail,
                grantType, userLogin.UserId);
        }
        #endregion

        #region Internal
        private static void TestCloudSync(this DBBase dexie)
        {
            if (!dexie.CloudSync)
            {
                throw new InvalidOperationException("Cloud Sync not enabled, call 'Create' with DexieCloudOptions.");
            }
        }
        #endregion
    }
}

File: DexieNET/yarn/src/dexieNETCloud.ts
/*
dexieNET.js

Copyright(c) 2023 Bernhard Straub

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

'DexieNET' used with permission of David Fahlander 
*/

import { getTiedRealmId } from "dexie-cloud-addon";
import { DB } from "./dexieNETBase";
import { DexieCloudSchema } from "dexie-cloud-common";
import { DexieCloudOptions } from "dexie-cloud-addon/dist/modern/DexieCloudOptions";

export function ConfigureCloud(db: DB, cloudOptions: DexieCloudOptions, dotnetRef: any): string | null {
    try {
        if (dotnetRef !== null)
        {
            cloudOptions.fetchTokens = async (tokenParams) => {
                return await dotnetRef.invokeMethodAsync('FetchTokens', tokenParams);
            }
        }
        db.cloud.configure(cloudOptions);
    }
    catch (err) {
        return err.message
    }

    return null;
}

export function AddOnVersion(db: DB): string {
    return db.cloud.version;
}

export function CurrentUserId(db: DB): string {
    return db.cloud.currentUserId;
}

export function Options(db: DB): any {
    return db.cloud.options;
}

export function Schema(db: DB): DexieCloudSchema | null {
    return db.cloud.schema;
}

export function UsingServiceWorker(db: DB): boolean | undefined {
    return db.cloud.usingServiceWorker;
}

export async function UserLogin(db: DB, email: string, grantType: "demo" | "otp" | undefined, userId?: string): Promise<string | undefined> {

    let httpError: string | undefined = undefined;

    try {
        await db.cloud.login({ email: email, userId: userId, grant_type: grantType });
    }
    catch (e) {
        httpError = e.message;
    }

    return httpError;
}

// sync with DexieNETCloudSync.cs record Sync 
interface CloudSyncOptions {
    wait: boolean;
    purpose: number;
}

export async function Sync(db: DB, sync: CloudSyncOptions ): Promise<void> {
    await db.cloud.sync({purpose: sync.purpose == 0 ? "push" : "pull", wait: sync.wait });
}

export async function Logout(db: DB, force?: boolean | undefined): Promise<void> {
    await db.cloud.logout({ force: force });
}

export function GetTiedRealmID(id: string): string {
    return getTiedRealmId(id);
}

export async function AcceptInviteMember(db: DB, id: string): Promise<void> {
    await db.members.update(id, { accepted: new Date(), rejected: undefined })
}

export async function RejectInviteMember(db: DB, id: string): Promise<void> {
    await db.members.update(id, { rejected: new Date(), accepted: undefined })
}

export async function ClearInviteMember(db: DB, id: string): Promise<void> {
    await db.members.update(id, { rejected: undefined, accepted: undefined })
}